core/src/cgroup.rs

```diff
@@ -185,16 +188,30 @@ impl CGroup {
         // Calling kill will also kill all sub cgroup processes
         self.kill()?;
 
-        // Obtain all sub cgroups
-        let dirs: Vec<PathBuf> = fs::read_dir(&self.path)?
-            .filter(|entry| entry.as_ref().unwrap().metadata().unwrap().is_dir())
-            .map(|entry| entry.unwrap().path())
-            .collect();
-
-        for dir in dirs {
-            fs::remove_dir(dir)?;
+        // Delete all cgroups from deepest to highest.
+        // It is necessary to delete the outer-most directories first, because
+        // a non-empty directory may not be deleted.
+        trace!("Calling remove on {}", &self.path.display());
+        for d in WalkDir::new(&self.path)
+            .into_iter()
+            .flatten()
+            .filter(|e| e.file_type().is_dir())
+            .sorted_by(|a, b| a.depth().cmp(&b.depth()).reverse())
+        {
+            trace!("Removing cgroup {}", &d.path().display());
+
+            // Re-try on every EBUSY
+            loop {
+                match fs::remove_dir(d.path()) {
+                    Err(e) => {
```

Comment:
I guess the real question is why is the directory still busy, and how can we fix that.

Reply:
I agree! The most weird thing remains however, that this does not occur (for my machine) when running the hypervisor with super user privileges...

Reply:
Kind of the same question as https://github.com/aeronautical-informatics/apex-linux/pull/22#discussion_r1058656617